### PR TITLE
add TSN

### PR DESCRIPTION
**Pull request by WangYizhang01**

### CODE REVIEW

File: TSN/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# TSN
+
+This repo provides one method for Person Action Recognition task.
+
+## Get started
+Prerequisites
+pip3 install -r requirements.txt
```
Comment (Ldpe2G) [inline TSN/README.md:7]:
```bash
pip3 install -r requirements.txt
```

File: TSN/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+## Data preparation
+We use Kinetics400 to train and evaluate our models. Please refer to (https://github.com/open-mmlab/mmaction/tree/master/data_tools/kinetics400) for data prepration.
+download mini_dataset here for evaluate the model.
```
Comment (Ldpe2G) [inline TSN/README.md:11]: Download the mini_dataset to evaluate the model.

File: TSN/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# TSN
+
+This repo provides one method for Person Action Recognition task.
+
+## Get started
+Prerequisites
+pip3 install -r requirements.txt
+
+## Data preparation
+We use Kinetics400 to train and evaluate our models. Please refer to (https://github.com/open-mmlab/mmaction/tree/master/data_tools/kinetics400) for data prepration.
```
Comment (Ldpe2G) [inline TSN/README.md:10]: We train and evaluate our models on Kinetics400 dataset. Please refer to (https://github.com/open-mmlab/mmaction/tree/master/data_tools/kinetics400) for data prepration.

File: TSN/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+##Train a model
+To train a model, run sh train.sh
+
+Our model is trained using ResNet50 as backbones
+##Evaluate a model
+
+To evaluate a model, run sh infer.sh
+
+## accuracy
```
Comment (Ldpe2G) [inline TSN/README.md:25]: Accuracy

File: TSN/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+```
+
+##Train a model
+To train a model, run sh train.sh
+
+Our model is trained using ResNet50 as backbones
+##Evaluate a model
+
+To evaluate a model, run sh infer.sh
+
```
Comment (Ldpe2G) [inline TSN/README.md:24]: ## Training
```bash
bash train.sh
```
Our model is trained using ResNet50 as backbones.

## Evaluating
```bash
bash infer.sh
```

File: TSN/test_recognizer.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
+import argparse
+import time
+import numpy as np
+
+# from tsn.datasets.video_dataset import VideoDataset
```
Comment (Ldpe2G) [inline TSN/test_recognizer.py:5]: 删除

File: TSN/test_recognizer.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        results.append(result)
+
+        toc = time.time()
+        proc_time_pool = proc_time_pool + toc - tac
+
+        tic = toc
+    return results
+
+    # print(len(data_loader))
+    #
```
Comment (Ldpe2G) [inline TSN/test_recognizer.py:139]: 脚本中无用的注释代码都删掉

File: TSN/train_recognizer.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        "--data_dir",
+        default="/home/liling/work/oneflow/mmaction/data",
+        help="data file path",
+    )
+    parser.add_argument(
+        "--pretrained",
+        default="/home/liling/work/oneflow/TSN/weights/resnet50_imagenet_pretrain_model",
+        help="test videos prefix path",
+    )
+    # parser.add_argument('--normvalue', default={'mean': [123.675, 116.28, 103.53], 'std': [58.395, 57.12, 57.375], 'to_rgb': True},
```
Comment (Ldpe2G) [inline TSN/train_recognizer.py:32]: 删除

File: TSN/tsn/datasets/dataset.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    def label(self):
+        return int(self._data[2])
+
+
+class TSNDataSet(object):
+    def __init__(
+        self,
+        root_path,
+        list_file,
+        video_dir="/home/liling/work/oneflow/mmaction/data/kinetics400/rawframes_val",
```
Comment (Ldpe2G) [inline TSN/tsn/datasets/dataset.py:31]: 不要出现这种特定的路径

File: TSN/tsn/models/TSN.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+        x = self.spatial_temporal_module(x)
+        x = x.reshape((-1, num_seg) + tuple(x.shape[1:]))
+
+        x = self.segmental_consensus(x)
+        x = x.squeeze(1)
+        losses = dict()
+
+        cls_score = self.cls_head(x)
+        # gt_label = gt_label.squeeze()
```
Comment (Ldpe2G) [inline TSN/tsn/models/TSN.py:67]: 脚本中有大量的注释代码，都检查下删掉吧

File: TSN/tsn/models/resnet50.py [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+
+        x = self.layer1(x)
+        x = self.layer2(x)
+        x = self.layer3(x)
+        x = self.layer4(x)
+
+        # x = self.avgpool(x)
+        # x = flow.flatten(x, 1)
+        # x = self.fc(x)
+
```
Comment (Ldpe2G) [inline TSN/tsn/models/resnet50.py:257]: 删

File: TSN/tsn/utils/checkpoint.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        state_dict = checkpoint["state_dict"]
+    else:
+        state_dict = checkpoint
+
+    model.load_state_dict(state_dict)
+
+    # parameters=model.state_dict()
+    # print(parameters)
+    # for key, value in parameters.items():
+    #     print(value)
```
Comment (Ldpe2G) [inline TSN/tsn/utils/checkpoint.py:32]: 以上删

File: TSN/train.sh [hunk 13]
```diff
@@ -0,0 +1,12 @@
+set -aux
+
+PRETRAIN_MODEL_PATH="resnet50_imagenet_pretrain_model"
+DATASET_PATH="../data"
```
Comment (Ldpe2G) [inline TSN/train.sh:4]: 这里不应该用相对路径，数据集根目录应该和 `train.sh` 脚本同级才对，这里再检查一下。
或者把数据集下载的命令也放到脚本里，首先检查数据集目录，不存在就下载解压。

Comment (WangYizhang01) [reply TSN/train.sh]: 以上review均已修改

File: TSN/train.sh [hunk 14]
```diff
@@ -0,0 +1,20 @@
+set -aux
+
+PRETRAIN_MODEL_PATH="resnet50_imagenet_pretrain_model"
+DATASET_PATH="./data"
+SAVE_PATH="./save_path"
+
+if [ ! -d "$DATASET_PATH" ]; then
+    mkdir data
```
Comment (Ldpe2G) [inline TSN/train.sh:8]: 这里不用下载数据集吗

File: TSN/README.md [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+```
+
+## Data preparation
+
+We train and evaluate our models on Kinetics400 dataset. Please refer to (https://github.com/open-mmlab/mmaction/tree/master/data_tools/kinetics400) for data prepration.
+
+Download the mini_dataset to evaluate the model.
+
+```bash
+wget https://oneflow-public.oss-cn-beijing.aliyuncs.com/datasets/TSN/kinetics400.zip
```
Comment (Ldpe2G) [inline TSN/README.md:18]: 感觉这里就不用单独列出数据集下载链接了，因为运行脚本就会自行下载

File: TSN/train.sh (hunk 14 above)
Comment (WangYizhang01) [reply TSN/train.sh:8]: 这个训练数据集是需要用户根据github介绍准备的，放在./data中，数据集太大了没法上传。

Comment (WangYizhang01) [reply TSN/train.sh:8]: REDEME里 data perpetration里面有

File: TSN/README.md (hunk 15 above)
Comment (WangYizhang01) [reply TSN/README.md:18]: 好的

File: TSN/train.sh (hunk 14 above)
Comment (Ldpe2G) [reply TSN/train.sh:8]: 这里再加个简单的提示吧

File: TSN/tsn/utils/checkpoint.py (hunk 12 above)
Comment (Ldpe2G) [reply TSN/tsn/utils/checkpoint.py:32]: 这里没改

File: TSN/tsn/models/resnet50.py (hunk 11 above)
Comment (Ldpe2G) [reply TSN/tsn/models/resnet50.py:257]: 这里没改

File: TSN/tsn/models/resnet50.py [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+        )
+        self.layer3 = self._make_layer(
+            block, 256, layers[2], stride=2, dilate=replace_stride_with_dilation[1]
+        )
+        self.layer4 = self._make_layer(
+            block, 512, layers[3], stride=2, dilate=replace_stride_with_dilation[2]
+        )
+        # self.avgpool = nn.AvgPool2d((7, 7))
+        # self.fc = nn.Linear(512 * block.expansion, num_classes)
+
```
Comment (Ldpe2G) [inline TSN/tsn/models/resnet50.py:181]: 上面没删

File: TSN/tsn/models/cls_head.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+                        .unsqueeze(-1),
+                        "bias": self.fc_cls.bias,
+                    }
+                )
+            class_map = self.new_cls(x)
+            return class_map
+
+    def loss(self, cls_score, labels):
+        # losses = dict()
+        # losses['loss_cls'] = nn.CrossEntropyLoss(cls_score, labels)
```
Comment (Ldpe2G) [inline TSN/tsn/models/cls_head.py:88]: 没删

File: TSN/tsn/models/cls_head.py [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+            assert x.shape[4] == self.spatial_feature_size
+            if self.with_avg_pool:
+                x = self.avg_pool(x)
+            if self.dropout is not None:
+                x = self.dropout(x)
+
+            shape = (x.shape[0], x.shape[1] * x.shape[2] * x.shape[3])
+            x = x.reshape(shape)
+            # print(x)
+            # x = x.view(x.size(0), -1)
```
Comment (Ldpe2G) [inline TSN/tsn/models/cls_head.py:63]: 这里删

File: TSN/tsn/models/TSN.py [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+        x = self.extract_feat(img_group)
+        # print(x)
+        x = self.spatial_temporal_module(x)
+        x = x.reshape((-1, num_seg) + tuple(x.shape[1:]))
+        x = self.segmental_consensus(x)
+        x = x.squeeze(1)
+        x = self.cls_head(x)
+        # print(x.shape)
+        # print(x)
```
Comment (Ldpe2G) [inline TSN/tsn/models/TSN.py:98]: 这里删

File: TSN/train.sh (hunk 14 above)
Comment (WangYizhang01) [reply TSN/train.sh:8]: 好的

File: TSN/tsn/models/TSN.py (hunk 10 above)
Comment (Ldpe2G) [reply TSN/tsn/models/TSN.py:67]: 这里没改

File: TSN/tsn/datasets/dataset.py (hunk 9 above)
Comment (Ldpe2G) [reply TSN/tsn/datasets/dataset.py:31]: 这里没改

**Review by Ldpe2G — APPROVED** [review]
